### ISSUE TITLE
Document the use of a `gen/main.go`, as a wrapper to `goagen`

### ISSUE THREAD

**Issue by abourget** [goadesign/goa#341]
#### The problem

Tools like `govendor` analyse the code in the repo to determine what needs to be vendored in.  Since  `goagen` generates codes on-the-fly, there are no _imports_ left in the repo for govendor to discover.

I want my projects to be fully contained with all dependencies, so my team checking out, can tweak the design, update the code, without fear its going to use a newer goa version, or something that happens to be installed on their machine.

A lesser problem is speed of re-generation. Since goagen generates a new `main.go` on-the-fly for each invocation, it needs to compile and run all those invocations, some are longer than other, but it can be sped up by having parsing done only once, and run all generators in a swift.
#### The proposition

I propose offering users this solution to setup their environment, instead of the inevitable `bash` script / `Makefile` that will be invoking `goagen` many times:

In your project's `gen/main.go`:

```
package main

import (
    _ "github.com/path/to/your/design"

    "github.com/goadesign/goa/design"
    "github.com/goadesign/goa/dslengine"
    "github.com/goadesign/goa/goagen/gen_app"
    "github.com/goadesign/goa/goagen/gen_swagger"
)

// Call with: "go run gen/main.go" from project root

func main() {
    dslengine.FailOnError(dslengine.Errors)
    dslengine.FailOnError(dslengine.Run())

    genapp.TargetPackage = "app"
    dslengine.PrintFilesOrFail(new(genapp.Generator).Generate(design.Design))

    dslengine.PrintFilesOrFail(new(genswagger.Generator).Generate(design.Design))

    // run custom generators...
}
```

This has several benefits:
- It's a lot faster, as the DSL is compiled and parsed only once.
- It would make the settings and generators used versionable, along with what is usually in the compiled `goagen` binary.
- It would _greatly_ improves the vendoring support: having code in the repo allows `govendor` to discover all dependencies, whereas `goagen` produces code that is compiled on-the-fly, so `govendor` cannot discover it. And it's a pain to add manually.
- Custom generators would not be aliens, like they seem in the current `goagen` invocations, while others have the special treatment of being "internal".
- It would decouple the Generator's command-line flags support from their core generation capabilities. Right now, things are quite intertwined.
- It simplifies the re-generation after modification of your `design`, with a simple call to `go run gen/main.go`. No need to pass on the design path: it won't change anyway.
- It introduces our users to the goa code up front, making it discoverable and approachable, because very simple.

We could still simplify the proposed in these ways:
- By breaking out of the `Generator` interface, and simply making the `Generate()` function a convention, we could pass the "TargetPackage" to the `genapp` Generator as a parameter. Taking another line out
- By getting rid of the commandline things in the different `gen_` packages, we could eventually invoke the generators with their own signature, like `genapp.Generate(design.Design)` and `genswagger.Generate()`.

``` go
    dslengine.PrintFilesOrFail(genapp.Generate(design.Design, "app"))

    // Instead of: 
    // genapp.TargetPackage = "app"
    // dslengine.PrintFilesOrFail(new(genapp.Generator).Generate(design.Design))
```
- We could find a way to kill the first line in `main()` too..

**Comment by raphael** [comment]
Wow there's a lot here :) A few comments:

1. Decoupling the generators from the command line parsing makes a lot of sense. Things are the way they are today not by design just as a result of how they evolved.
2. It also makes sense to make it easier to write tools the way this proposal suggests. That is rewrite the goagen package so that it's easier to use it as such rather than just through the tool.
3. That being said I do believe that there is a lot of benefit to having the one tool. It would be good to avoid having all tools become their own little snowflake.

In other words cleaning up the goagen package to make it independent of the tool seems like a good thing to do that could enable interesting use for the package.

However I don't think that usage pattern should become the default for multiple reasons:

1. Writing a plugin as things are today is trivial, implement a package that exposes `Generate` - done (again modulo cleaning up how the command line is parsed and the flags used).
2. There is one tool. You learn that one tool and now can generate all kinds of outputs. It provides for a consistent experience.
3. Having one tool also gives the opportunity for adding more discoverability. For example If we have a place where plugins can be registered then `goagen` could discover them, list them document them etc.

At the end of the day the vision is that `goagen` is a central piece of `goa`. A lot of the benefits you list can be achieved with that still being true though. I would suggest breaking this issue down into smaller ones tackling each issue keeping that in mind so we can have more concrete discussions. Thank you for your input!

**Comment by abourget** [comment]
I updated this proposition to reflect a bit more of our discussion.  I think it might just be useful to add an example somewhere for those in need of such a thing.

**Comment by raphael** [comment]
Closing this old issue because no activity, feel free to open a new issue.
